cpp/include/raft/sparse/matrix/detail/preprocessing.cuh

```diff
@@ -0,0 +1,463 @@
+ */
+template <typename T1, typename T2, typename IdxT>
+void get_uniques_counts(raft::resources& handle,
+                        raft::device_vector_view<T1, IdxT> sort_vector,
+                        raft::device_vector_view<T1, IdxT> secondary_vector,
+                        raft::device_vector_view<T2, IdxT> data,
+                        raft::device_vector_view<T2, IdxT> itr_vals,
+                        raft::device_vector_view<T1, IdxT> keys_out,
+                        raft::device_vector_view<T2, IdxT> counts_out)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+
+  raft::sparse::op::coo_sort(sort_vector.size(),
+                             secondary_vector.size(),
+                             data.size(),
+                             sort_vector.data_handle(),
+                             secondary_vector.data_handle(),
+                             data.data_handle(),
+                             stream);
+
+  thrust::reduce_by_key(raft::resource::get_thrust_policy(handle),
+                        sort_vector.data_handle(),
+                        sort_vector.data_handle() + sort_vector.size(),
+                        itr_vals.data_handle(),
+                        keys_out.data_handle(),
+                        counts_out.data_handle());
+}
+
+/**
+ * @brief Compute cumulative sum for each unique value in the origin array
+ * @param handle: raft resource handle
+ * @param origin: Input array that has values to use for computation
+ * @param keys: Output array that has keys, should be the size of unique
+ * @param counts: Output array that contains the computed counts
+ * @param results: Output array that scatters the counts to origin value positions. Same size as
+ * origin array.
+ */
+template <typename T1, typename T2, typename IdxT>
+void create_mapped_vector(raft::resources& handle,
+                          const raft::device_vector_view<T1, IdxT> origin,
+                          const raft::device_vector_view<T1, IdxT> keys,
+                          const raft::device_vector_view<T2, IdxT> counts,
+                          raft::device_vector_view<T2, IdxT> result)
+{
+  cudaStream_t stream = raft::resource::get_cuda_stream(handle);
+  auto host_keys      = raft::make_host_vector<T1, IdxT>(handle, keys.size());
+
+  raft::copy(host_keys.data_handle(), keys.data_handle(), keys.size(), stream);
+  raft::linalg::map(handle, result, raft::cast_op<T2>{}, raft::make_const_mdspan(origin));
+  int new_key_size = host_keys(host_keys.size() - 1) + 1;
+  auto origin_map  = raft::make_device_vector<T2, IdxT>(handle, new_key_size);
+
+  thrust::scatter(raft::resource::get_thrust_policy(handle),
+                  counts.data_handle(),
+                  counts.data_handle() + counts.size(),
+                  keys.data_handle(),
+                  origin_map.data_handle());
+  thrust::for_each(raft::resource::get_thrust_policy(handle),
```

Comment:
Is the goal here to do an elementwise op over an input and write to the output? We should be able to use RAFT for this.